### ISSUE TITLE
CUDA register pressure due to unnecessarily wide pointers.

### ISSUE THREAD

**Issue by mcourteaux** [halide/Halide#7884]
Examining performance of my CUDA kernel generated by Halide, I saw that in the SASS code, integer instructions with a `.WIDE` suffix cause 2 registers to be used at once, which makes sense as the `.WIDE` suffix refers to 64-bit.
![image](https://github.com/halide/Halide/assets/845012/2fe3556b-a6bc-4096-97a2-d34f9d1dc018)

Trying to get the register usage down, this is an obvious candidate to replace with 16-bit or 32-bit arithmetic. I was doing some research online, and this seems supported: https://docs.nvidia.com/cuda/archive/10.0/parallel-thread-execution/index.html#using-addresses-arrays-and-vectors

I can see that `promote_indices()` is set to to `return false` for PTX devices CodeGen, so that suggests that the indices that I'm using are already 64-bit in the Expr coming from the Generator's pipeline code. However, the official documentation about `Var` says:

> It is just a name, and can be reused in places where no name conflict will occur. It can be used in the left-hand-side of a function definition, or as an [Expr](https://halide-lang.org/docs/struct_halide_1_1_expr.html). As an [Expr](https://halide-lang.org/docs/struct_halide_1_1_expr.html), it always has type Int(32).

So somehow, very straightfoward usage of Halide causes the indices to go up to 64-bit, unless I'm wrong and the indices stay 32-bit, and it's the base-pointer that is 64-bit and causes everything to go to 64-bit.

---
In general, I don't think LLVM ever produces `[%r45]` for example, it always uses the wide registers (e.g., `%rd45`). It for sure didn't for my kernels in this pipeline.
For example:
![image](https://github.com/halide/Halide/assets/845012/6afca0ef-4484-47df-be5e-bb929a7eb962)

Now, to zoom a little bit in into the base-pointer being 64-bit idea, I don't think that's valid, because for example, I'm using `GPUShared` storage, which in CUDA has very low addresses. At times, you can see that it compiles hardcoded loads like: `ld.shared.f32 [0x100]` for example. So for sure it should be possible to keep addressing arithmetic to 32-bit (or even 16-bit).

---

Is this possible today with Halide, or should things change?

**Comment by abadams** [comment]
I suspect what's happening is that pointers are 64-bit, and indices are 32-bit, but indices are in terms of elements, not bytes, so we don't always know if 32-bit address math would be enough. When the pointers are 64-bit, in the llvm API we need to cast the index to 64-bit before a GEP node that does the pointer math. When llvm lowers the GEP node it multiplies by the size of the type, still in 64-bit. On most targets, this multiplication is free because it's part of the addressing mode. That does not seem to be the case in PTX - there's an explicit multiplication by four.

If we statically know the allocation is < 2gb (e.g. constant-size allocations, shared memory on GPUs), we could perhaps do the address math ourselves in 32-bit and then cast the pointer back to the element type before the load instead of using a GEP node. I'm not sure if that would mess up other LLVM optimizations though. I think the conservative thing to do would be to virtualize CodeGen_LLVM::codegen_buffer_pointer and then implement it differently in the PTX backend.

**Comment by mcourteaux** [comment]
> If we statically know the allocation is < 2gb (e.g. constant-size allocations).

I don't think this will even be possible, right? Constant size doesn't necessarily mean that the base address will be within 32-bit range, I would assume?

For sure, it should be possible for `local` and `shared` memory.

**Comment by abadams** [comment]
We can statically know the index, expressed as a byte offset, is expressible in 32 bits, and this might enable cheaper addressing modes.

**Comment by mcourteaux** [comment]
I don't follow. The base pointer being 64-bit in case of global memory (e.g. 24GB of VRAM on an A100), seems to be the problem. Even if the index is compile-time known to be small, we won't know the base-pointer, and you still need 64-bit registers either way.

Maybe I'm talking about something else.

**Comment by abadams** [comment]
Never mind, I misread the cuda docs and thought there was an addressing mode that takes a 64-bit base pointer and a 32-bit offset, but it looks like the offset has to be constant to use that mode.

**Comment by mcourteaux** [comment]
I believe you are referring to: https://docs.nvidia.com/cuda/archive/10.0/parallel-thread-execution/index.html#addresses-as-operands

Yes, as far as I can tell, there is indeed no `reg + reg` addressing mode.

**Comment by mcourteaux** [comment]
I was toying around with this a little bit, and reading upon LLVM address spaces and the GEP instruction. As far as I understand, I believe it would suffice to: (1) delete the promotion code that is sprinkled left and right in the `CodeGen_LLVM::codegen_buffer_pointer` functions, and (2) make sure somehow that the `llvm::Value*` for `base_address` of the buffers is by default of the right type in case of `local` or `shared` memory buffers (i.e.: i32_t, I think?).

I am not at all very familiar yet with LLVM codegen, nor with how Halide internally manages the buffers (it seems to be by-name, and I haven't dug through the code enough yet). I'm mainly commenting this, because I wanted to check if I was on the right path to try to implement a fix for this, as my idea is different from what you initially commented, @abadams. What I was wondering is if deleting the promotion code is a sane idea? Wouldn't LLVM just by default generate the right code, regardless of when the `llvm::Value *base_address` is 32-bit or 64-bit?

I would need more toying around I guess to figure this out.

**Comment by abadams** [comment]
The promotion code was very useful to get llvm to use pointer marching instead of explicit indexing, i.e. turning this sort of code:

```
for (int i =0; i < 100; i++) {
  ... array[i] ...
}
```

into this sort of code:

```
float *a = array;
for (int i =0; i < 100; i++) {
  ... *a++ ...
}
```

It didn't want to do that with 32-bit indices at the time the promotion code was written (which, to be fair, was a long time ago).

Maybe it's not necessary anymore, but we'd have to test that pretty carefully. I think it would be safer to just replace codegen_buffer_pointer for ptx.

**Comment by mcourteaux** [comment]
Okay, thanks for the helpful background. I'll give it another go sometime soon. :slightly_smiling_face:

**Comment by mcourteaux** [comment]
Attempt with inline assembly indeed seems viable.

```ll
%ptrC = mul i32 4, %id
call void asm sideeffect "st.shared.f32 [$0], $1;", "r,f"(i32 %ptrC, float %valC)
```

produces:

```c
	shl.b32 	%r1, %r2, 2;
	// begin inline asm
	st.shared.f32 [%r1], %f1;
	// end inline asm
```
     
Which is definitely a good start.

This approach unfortunately forces us to rewrite stores and loads, rather then just replacing the `GEP` (i.e.: `codegen_buffer_pointer`). That would require us to create `virtual` functions `codegen_load()` and `codegen_store()` in `CodeGen_LLVM`, I guess.

**Comment by mcourteaux** [comment]
I'm surprised. The NVIDIA compiler that takes in the PTX to produce the SASS is smarter than LLVM and seems to convert a lot of those 64-bit addressing to shared and local memory to 32-bit addressing:

![image](https://github.com/halide/Halide/assets/845012/0988c233-d6d9-4992-8098-cab9b95c5b7c)

the `LDG` (load global, (1) in screenshot) uses `R25.64` ((2) in screenshot, which is the concatenation of `R25` and `R26` to get a 64-bit register), whereas the `LDS` (load shared) uses `R39`.

**Comment by abadams** [comment]
Does that mean this issue is moot?

**Comment by mcourteaux** [comment]
I'll keep investigating a little bit over the coming days. I stopped working on that kernel, as I was trying to get Halide to produce the right PTX, but maybe the NVIDIA driver takes care of it reliably. I'll check keep an eye on this, and see if there are cases where it doesn't do this optimization.

Generally, I don't like that we rely on an optimization in the driver, rather than just producing the correct PTX.

**Comment by mcourteaux** [comment]
![image](https://github.com/halide/Halide/assets/845012/1b808d38-61df-4e53-92d1-4e5d1339d720)

Here is another example of the NVIDIA driver not optimizing it: `%r247` (32-bit) gets multiplied by `sizeof(float)` and turns into a 64-bit number `%rd141`, which is used to store into shared memory.
